pkg/auth/policy_store.go

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2024 The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package auth
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang-collections/collections/set"
+
+	"kmesh.net/kmesh/api/v2/workloadapi/security"
+)
+
+type policyStore struct {
+	// byKey maintains a mapping of ns/name to policy
+	byKey map[string]authPolicy
+
+	// byNamespace maintains a mapping of namespace (or "" for global) to policy names
+	byNamespace map[string]*set.Set
+
+	rwLock sync.RWMutex
+}
+
+func newPolicystore() *policyStore {
+	return &policyStore{
+		byKey:       make(map[string]authPolicy),
+		byNamespace: make(map[string]*set.Set),
+	}
+}
+
+func (ps *policyStore) updatePolicy(auth *security.Authorization) error {
+	if auth == nil {
+		return nil
+	}
+
+	authPolicy := authPolicy{
+		auth,
+	}
+	key := authPolicy.Key()
+
+	var ns string
+	switch authPolicy.GetScope() {
+	case security.Scope_GLOBAL:
+		ns = ""
+	case security.Scope_NAMESPACE:
+		ns = authPolicy.GetNamespace()
+	case security.Scope_WORKLOAD_SELECTOR:
+		// do nothing
+	default:
+		return fmt.Errorf("invalid scope %v of authorization policy", authPolicy.GetScope())
+	}
+
+	ps.rwLock.Lock()
+	defer ps.rwLock.Unlock()
+
+	if s, ok := ps.byNamespace[ns]; !ok {
+		ps.byNamespace[ns] = set.New(key)
+	} else {
+		s.Insert(key)
+	}
+
+	ps.byKey[key] = authPolicy
+	return nil
+}
+
+func (ps *policyStore) removePolicy(policyKey string) error {
+	ps.rwLock.Lock()
+	defer ps.rwLock.Unlock()
+
+	authPolicy, ok := ps.byKey[policyKey]
+	if !ok {
+		log.Warnf("Auth policy key %s does not exist in byKey", policyKey)
+		return nil
+	}
+
+	var ns string
+	switch authPolicy.Scope {
+	case security.Scope_GLOBAL:
+		ns = ""
+	case security.Scope_NAMESPACE:
+		ns = authPolicy.GetNamespace()
+	case security.Scope_WORKLOAD_SELECTOR:
+		// do nothing
+	default:
+		return fmt.Errorf("invalid scope %v of authorization policy", authPolicy.GetScope())
```

Comment:
we donot need these two cases, it can never reach here. We have guaranteed in `updatePolicy`

Reply:
done